ast/src/parsed/pattern.rs

```diff
@@ -0,0 +1,411 @@
+                        new_patterns.push(unspecialized);
+                        new_patterns.extend_from_slice(remaining_part);
+                        Some(PatternTuple {
+                            patterns: new_patterns,
+                        })
+                    } else {
+                        None
+                    }
+                }
+                Pattern::Array(cons_patterns) => {
+                    let length = cons_patterns.len();
+                    if tuple.patterns.len() >= length {
+                        let (to_unspecialize, remaining_part) = tuple.patterns.split_at(length);
+                        let unspecialized = Pattern::Array(to_unspecialize.to_vec());
+
+                        let mut new_patterns = Vec::with_capacity(remaining_part.len() + 1);
+                        new_patterns.push(unspecialized);
+                        new_patterns.extend_from_slice(remaining_part);
+                        Some(PatternTuple {
+                            patterns: new_patterns,
+                        })
+                    } else {
+                        None
+                    }
+                }
+
+                Pattern::Enum(path1, _) => Some(PatternTuple {
+                    patterns: vec![Pattern::Enum(path1.clone(), Some(tuple.patterns))],
+                }),
+                _ => None,
+            }
+        }
+    }
+}
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub enum Pattern {
+    CatchAll, // "_", matches a single value
+    Ellipsis, // "..", matches a series of values, only valid inside array patterns
+    #[schemars(skip)]
+    Number(BigInt),
+    String(String),
+    Tuple(Vec<Pattern>),
+    Array(Vec<Pattern>),
+    // A pattern that binds a variable. Variable references are parsed as
+    // Enum and are then re-mapped to Variable if they do not reference
+    // an enum variant.
+    Variable(String),
+    Enum(SymbolPath, Option<Vec<Pattern>>),
+}
+
+impl Pattern {
+    /// Returns an iterator over all variables in this pattern.
+    pub fn variables(&self) -> Box<dyn Iterator<Item = &String> + '_> {
+        match self {
+            Pattern::Variable(v) => Box::new(once(v)),
+            _ => Box::new(self.children().flat_map(|p| p.variables())),
+        }
+    }
+
+    /// Return true if the pattern is irrefutable, i.e. matches all possible values of its type.
+    pub fn is_irrefutable(&self) -> bool {
+        match self {
+            Pattern::Ellipsis => unreachable!(),
+            Pattern::CatchAll | Pattern::Variable(_) => true,
+            Pattern::Number(_) | Pattern::String(_) | Pattern::Enum(_, _) => false,
+            Pattern::Array(items) => {
+                // Only "[..]"" is irrefutable
+                items == &vec![Pattern::Ellipsis]
+            }
+            Pattern::Tuple(p) => p.iter().all(|p| p.is_irrefutable()),
+        }
```

Comment:
Can you add a TODO or something that this should also return true for enum variants that are the only ones in the enum?